### PR TITLE
Change updates demos/http3 so it uses `SSL_write_ex2()` together with

### PR DESCRIPTION
**Pull request by Sashan**
`SSL_WRITE_FLAG_CONCLUDE`. Both were introduced by PR #23343.

Change also does a minor tweak to Makefile so I can use env. variables to reach for nghttp3 on my system.

##### Checklist

[quic-pkts.tgz](https://github.com/openssl/openssl/files/14296137/quic-pkts.tgz)

I'm attaching .tgz file which contains three .pcap files:

- quic.pcap packet dumps gathered for code we have in master branch
- quic-fixed.pcap packet dumps gathered for code with my change applied
- quic-broken.pcap packet dumps gathered when running client which does not attempt to conclude connection

after examining packet dumps, it seems to me new code works as expected.
[quic-pkts.tgz](https://github.com/openssl/openssl/files/14296279/quic-pkts.tgz)

### CODE REVIEW

**Comment by kroeckx** [comment]
There is a typo in the commit message

**Comment by Sashan** [comment]
> There is a typo in the commit message

fixed, thanks.

File: demos/http3/ossl-nghttp3.c [hunk 1]
```diff
@@ -613,6 +614,12 @@ int OSSL_DEMO_H3_CONN_handle_events(OSSL_DEMO_H3_CONN *conn)
         if (ec == 0)
             break;
 
+        /*
+	 * we let SSL_write_ex2(3ossl) to conclude the stream for us (send FIN)
```
Comment (t8m) [inline demos/http3/ossl-nghttp3.c:618]: Nits: `We` and remove the `3ossl` or replace with `3` as I think the ossl is distibution-patch-specific thing.

**Review by t8m — CHANGES_REQUESTED** [review]
One nit

File: demos/http3/ossl-nghttp3.c [hunk 2]
```diff
@@ -567,6 +567,7 @@ int OSSL_DEMO_H3_CONN_handle_events(OSSL_DEMO_H3_CONN *conn)
     int ec, fin;
     size_t i, num_vecs, written, total_written, total_len;
     int64_t stream_id;
+    uint64_t flags;
```
Comment (shahsb) [inline demos/http3/ossl-nghttp3.c:570]: Please initialize this variable here.

Remember, it's always a good practice to initialize a variable with default values at the time of declaration.

**Review by shahsb — CHANGES_REQUESTED** [review]

File: demos/http3/ossl-nghttp3.c (hunk 2 above)
Comment (t8m) [reply demos/http3/ossl-nghttp3.c:570]: Why? This is not used uninitialized.

Comment (tom-cosgrove-arm) [reply demos/http3/ossl-nghttp3.c:570]: Agreed @t8m - we really don't want to require such a change, as it would lead to a lot of code churn. Note that none of the variables are assigned "default" values here, or in much of the code base.

Best practice would be to assign to each variable as it is declared, and have it live for as narrow a scope as possible.  That’s not possible with OpenSSL coding style (C90) unfortunately.

If variables are not initialised to a default value, some compilers can detect when they are used without prior assignment - this only has to be found by one compiler to be useful. Always assigning a default precludes any compiler from doing this. Further, a "default" value is pretty much guaranteed to be wrong some of the time. 

In summary, I’d prefer to keep to the general style here, which is not to set a default value for everything

File: demos/http3/ossl-nghttp3.c [hunk 3]
```diff
@@ -613,6 +614,12 @@ int OSSL_DEMO_H3_CONN_handle_events(OSSL_DEMO_H3_CONN *conn)
         if (ec == 0)
             break;
 
+        /*
+	 * we let SSL_write_ex2(3) to conclude the stream for us (send FIN)
```
Comment (tom-cosgrove-arm) [inline demos/http3/ossl-nghttp3.c:618]: Minor nit: "let" doesn't need "to" here

```suggestion
	 * Let SSL_write_ex2(3) conclude the stream for us (send FIN)
```

**Review by tom-cosgrove-arm — APPROVED** [review]
Happy to approve either with or without the nit in the comment.

I'd also suggest next time to have unrelated changes (`SSL_WRITE_FLAG_CONCLUDE` and `Makefile` changes) in different commits

**Comment by Sashan** [comment]
Hello,
On Mon, Feb 26, 2024 at 12:45:02AM -0800, Tom Cosgrove wrote:
>    @tom-cosgrove-arm approved this pull request.
> 
>    Happy to approve either with or without the nit in the comment.

    thanks for approving with nit.

> 
>    I'd also suggest next time to have unrelated changes
>    (SSL_WRITE_FLAG_CONCLUDE and Makefile changes) in different commits

    understood.


thanks and
regards
sashan

**Comment by t8m** [comment]
Merged to the master branch, thank you.
